File: lib/effs/tls_accept.c
// TLS server handshake: accepts one TCP connection (parked, like
// TCP.accept) then handshakes it on the helper thread (blocking, so the
// loop stays free). Answers Listener & Result with an SSL session box
// (TlsConn*), never a raw fd. OpenSSL rides dlopen, no link flags.
#include <dlfcn.h>
#include <fcntl.h>
#include <sys/socket.h>
#include <unistd.h>
#include <string.h>

#ifndef TLS_CONN_DEFINED
#define TLS_CONN_DEFINED
struct TlsConn {
  void* ssl;
  void* ctx;
  int   fd;
};
#endif

struct TlsServe {
  char* cert;
  u64   cert_len;
  char* key;
  u64   key_len;
  int   fd;
  u32   code;
  void* conn;
};

static void*          tls_s_h = NULL;
static pthread_mutex_t tls_s_gate = PTHREAD_MUTEX_INITIALIZER;
static int            tls_s_ready = 0;
static void* (*s_TLS_server_method)(void);
static void* (*s_SSL_CTX_new)(void*);
static void  (*s_SSL_CTX_free)(void*);
static int   (*s_SSL_CTX_use_certificate_file)(void*, const char*, int);
static int   (*s_SSL_CTX_use_PrivateKey_file)(void*, const char*, int);
static void  (*s_SSL_CTX_set_alpn_select_cb)(void*, void*, void*);
static void* (*s_SSL_new)(void*);
static int   (*s_SSL_set_fd)(void*, int);
static int   (*s_SSL_accept)(void*);
static void  (*s_SSL_get0_alpn_selected)(const void*, const unsigned char**, unsigned*);
static int   (*s_SSL_shutdown)(void*);
static void  (*s_SSL_free)(void*);

static u32 tls_s_init(void) {
  pthread_mutex_lock(&tls_s_gate);
  if (tls_s_ready) {
    pthread_mutex_unlock(&tls_s_gate);
    return 0;
  }
  tls_s_h = dlopen("libssl.so.3", RTLD_NOW);
  if (tls_s_h == NULL) {
    tls_s_h = dlopen("libssl.so", RTLD_NOW);
  }
  if (tls_s_h == NULL) {
    pthread_mutex_unlock(&tls_s_gate);
    return 616;
  }
#define TLS_S_SYM(n)                                    \
  s_##n = dlsym(tls_s_h, #n);                           \
  if (s_##n == NULL) {                                  \
    pthread_mutex_unlock(&tls_s_gate);                  \
    return 616;                                         \
  }
  TLS_S_SYM(TLS_server_method)
  TLS_S_SYM(SSL_CTX_new)
  TLS_S_SYM(SSL_CTX_free)
  TLS_S_SYM(SSL_CTX_use_certificate_file)
  TLS_S_SYM(SSL_CTX_use_PrivateKey_file)
  TLS_S_SYM(SSL_CTX_set_alpn_select_cb)
  TLS_S_SYM(SSL_new)
  TLS_S_SYM(SSL_set_fd)
  TLS_S_SYM(SSL_accept)
  TLS_S_SYM(SSL_get0_alpn_selected)
  TLS_S_SYM(SSL_shutdown)
  TLS_S_SYM(SSL_free)
#undef TLS_S_SYM
  tls_s_ready = 1;
  pthread_mutex_unlock(&tls_s_gate);
  return 0;
}

static const char* tls_s_msg(u32 code) {
  switch (code) {
    case 611: return "bad input";
    case 613: return "handshake failed";
    case 615: return "alpn mismatch";
    case 616: return "tls error";
    default:  return "accept failed";
  }
}

// ALPN select: prefer h2 when the client offers it.
static int tls_alpn_cb(void* ssl, const unsigned char** out,
  unsigned char* outlen, const unsigned char* in, unsigned inlen, void* arg) {
  (void)ssl;
  (void)arg;
  unsigned i = 0;
  while (i + 1 < inlen) {
    unsigned l = in[i];
    if (i + 1 + l > inlen) {
      break;
    }
    if (l == 2 && in[i + 1] == 'h' && in[i + 2] == '2') {
      *out = in + i + 1;
      *outlen = 2;
      return 0;
    }
    i += 1 + l;
  }
  return 3;
}

static void tls_serve_call(IoWork* w) {
  struct TlsServe* s = (struct TlsServe*)w->data;
  s->code = 0;
  s->conn = NULL;
  u32 q = tls_s_init();
  if (q != 0) {
    s->code = q;
    return;
  }
  if (io_nul(s->cert, s->cert_len) || io_nul(s->key, s->key_len)) {
    s->code = 611;
    return;
  }
  // Blocking handshake: drop O_NONBLOCK for its duration.
  int fl = fcntl(s->fd, F_GETFL, 0);
  int nb = (fl >= 0 && (fl & O_NONBLOCK) != 0) ? 1 : 0;
  if (nb) {
    fcntl(s->fd, F_SETFL, fl & ~O_NONBLOCK);
  }
  void* ctx = s_SSL_CTX_new(s_TLS_server_method());
  if (ctx == NULL) {
    if (nb) fcntl(s->fd, F_SETFL, fl);
    close(s->fd);
    s->code = 613;
    return;
  }
  if (s_SSL_CTX_use_certificate_file(ctx, s->cert, 1) != 1
    || s_SSL_CTX_use_PrivateKey_file(ctx, s->key, 1) != 1) {
    s_SSL_CTX_free(ctx);
    if (nb) fcntl(s->fd, F_SETFL, fl);
    close(s->fd);
    s->code = 613;
    return;
  }
  s_SSL_CTX_set_alpn_select_cb(ctx, (void*)tls_alpn_cb, NULL);
  void* ssl = s_SSL_new(ctx);
  if (ssl == NULL) {
    s_SSL_CTX_free(ctx);
    if (nb) fcntl(s->fd, F_SETFL, fl);
    close(s->fd);
    s->code = 613;
    return;
  }
  s_SSL_set_fd(ssl, s->fd);
  int r = s_SSL_accept(ssl);
  if (nb) {
    int fl2 = fcntl(s->fd, F_GETFL, 0);
    if (fl2 >= 0) fcntl(s->fd, F_SETFL, fl2 | O_NONBLOCK);
  }
  if (r <= 0) {
    s_SSL_shutdown(ssl);
    s_SSL_free(ssl);
    s_SSL_CTX_free(ctx);
    close(s->fd);
    s->code = 613;
    return;
  }
  struct TlsConn* c = io_mem(malloc(sizeof(*c)));
  c->ssl = ssl;
  c->ctx = ctx;
  c->fd  = s->fd;
  s->conn = c;
  s->code = 0;
}

static Term tls_serve_pack(Env e, IoWork* w) {
  int lfd = (int)w->hand;
  struct TlsServe* s = (struct TlsServe*)w->data;
  Term r;
  if (s->code == 0) {
    r = io_done(e, io_hand((intptr_t)s->conn));
  } else {
    r = io_fail(e, s->code, tls_s_msg(s->code));
  }
  free(s->cert);
  free(s->key);
  free(s);
  return io_tup(e, io_hand(lfd), r);
}

static Term tls_accept_more2(Env e, IoWork* w);

Term tls_accept_raw_run(Env e, Term* f, IoWork* w) {
  w->hand = (intptr_t)io_hand_v(f[0]);
  // Stash cert/key now: accept_more may park before reading them, and
  // f is only valid on entry, so copy them into the serve box upfront
  // via a pre-box that accept_more picks up through w->size/word.
  // Simpler: copy here into heap and chain through w->data as a pair.
  // accept_more creates the serve box; patch it here instead: create
  // it now with fd=-1 (pending), then accept_more fills fd + works.
  struct TlsServe* s = io_mem(malloc(sizeof(*s)));
  s->cert = io_cstr(e, f[1], &s->cert_len);
  s->key = io_cstr(e, f[2], &s->key_len);
  s->fd = -1;
  s->code = 0;
  s->conn = NULL;
  w->data = (char*)s;
  // Custom entry: try accept now; on EAGAIN park; on fd, reuse s.
  int lfd = (int)w->hand;
  int got = accept(lfd, NULL, NULL);
  if (got >= 0 && fcntl(got, F_SETFL, fcntl(got, F_GETFL) | O_NONBLOCK) < 0) {
    close(got);
    got = -1;
  }
  if (got < 0) {
    if (errno == EAGAIN) {
      // Keep s for the wake: stash it in w->text slot via size hack.
      // w->data already holds s; the wake re-enters below.
      return io_wait_on(w, lfd, POLLIN, 0, tls_accept_more2);
    }
    Term r = io_fail(e, (u32)errno, NULL);
    free(s->cert);
    free(s->key);
    free(s);
    return io_tup(e, io_hand(lfd), r);
  }
  s->fd = got;
  return io_work(w, tls_serve_call, tls_serve_pack);
}

// Wake after listener park: s is already in w->data.
static Term tls_accept_more2(Env e, IoWork* w) {
  int lfd = (int)w->hand;
  struct TlsServe* s = (struct TlsServe*)w->data;
  int got = accept(lfd, NULL, NULL);
  if (got >= 0 && fcntl(got, F_SETFL, fcntl(got, F_GETFL) | O_NONBLOCK) < 0) {
    close(got);
    got = -1;
  }
  if (got < 0) {
    if (errno == EAGAIN) {
      return io_wait_on(w, lfd, POLLIN, 0, tls_accept_more2);
    }
    Term r = io_fail(e, (u32)errno, NULL);
    free(s->cert);
    free(s->key);
    free(s);
    return io_tup(e, io_hand(lfd), r);
  }
  s->fd = got;
  return io_work(w, tls_serve_call, tls_serve_pack);
}

static void __attribute__((constructor)) tls_accept_raw_use(void) {
  io_eff(CID_TLS_ACCEPT_RAW, tls_accept_raw_run, IO_READ);
}

File: lib/effs/tls_accept.js
// TLS server handshake: accepts one TCP connection (parked, like
// TCP.accept) then handshakes it with the cert/key files, ALPN h2 via
// a JSCallback select. Answers Listener & Result with a TlsConn box.
function tls_accept_raw(listener, cert, key, k) {
  const sys = io_sys();
  const bad = (code, text) => io_tup(listener, { $: "Fail", error: io_tup(code, text) });
  const lib = tls_lib();
  if (lib === null) {
    return bad(616, "tls error");
  }
  const lfd = listener;
  const go = () => {
    const fd = sys.accept(lfd, null, null);
    if (fd < 0) {
      const code = sys.errno();
      if (code === (sys.mac ? 35 : 11)) {
        io_park_on(lfd, false, k, go);
        return undefined;
      }
      return bad(code, "accept failed");
    }
    if (sys.fcntl(fd, 4, sys.fcntl(fd, 3, 0) | (sys.mac ? 4 : 0x800)) < 0) {
      const code = sys.errno();
      sys.close(fd);
      return bad(code, "accept failed");
    }
    return shake(fd);
  };
  const fail = (code, text, fd, ctx, ses, cb) => {
    try {
      if (ses) lib.SSL_free(ses);
      if (ctx) lib.SSL_CTX_free(ctx);
      if (cb) cb.close();
    } catch (e) {}
    sys.close(fd);
    return bad(code, text);
  };
  const shake = (fd) => {
    let ctx = null;
    let ses = null;
    let cb = null;
    try {
      ctx = lib.SSL_CTX_new(lib.TLS_server_method());
    } catch (e) {
      ctx = null;
    }
    if (!ctx) {
      return fail(613, "handshake failed", fd, null, null, null);
    }
    try {
      const cert_b = io_bytes(cert + "\0");
      const key_b = io_bytes(key + "\0");
      if (lib.SSL_CTX_use_certificate_file(ctx, lib.ptr(cert_b), 1) !== 1) {
        return fail(613, "handshake failed", fd, ctx, null, null);
      }
      if (lib.SSL_CTX_use_PrivateKey_file(ctx, lib.ptr(key_b), 1) !== 1) {
        return fail(613, "handshake failed", fd, ctx, null, null);
      }
      // ALPN select: SSL_select_next_proto does the negotiation and
      // writes the out params itself, so the callback only forwards
      // pointers (no JS writes into native memory). The server list
      // (h2) lives in a global slot so it outlives the handshake.
      if (globalThis.BEND_TLS_SRV_ALPN === undefined) {
        globalThis.BEND_TLS_SRV_ALPN = new Uint8Array([2, 104, 50]);
      }
      const srv = globalThis.BEND_TLS_SRV_ALPN;
      cb = new lib.JSCallback((s, out, outlen, inp, inlen, arg) => {
        try {
          // 1 = OPENSSL_NPN_NEGOTIATED -> SSL_TLSEXT_ERR_OK (0);
          // 2 = NO_OVERLAP -> NOACK (3).
          const rv = lib.SSL_select_next_proto(out, outlen, lib.ptr(srv), 3, inp, Number(inlen));
          return rv === 1 ? 0 : 3;
        } catch (e) {
          return 3;
        }
      }, { args: ["ptr", "ptr", "ptr", "ptr", "u32", "ptr"], returns: "i32" });
      lib.SSL_CTX_set_alpn_select_cb(ctx, cb.ptr, null);
      ses = lib.SSL_new(ctx);
      if (!ses) {
        return fail(613, "handshake failed", fd, ctx, null, cb);
      }
      lib.SSL_set_fd(ses, fd);
    } catch (e) {
      return fail(613, "handshake failed", fd, ctx, ses, cb);
    }
    const step = () => {
      let r = -1;
      try {
        r = lib.SSL_accept(ses);
      } catch (e) {
        return fail(613, "handshake failed", fd, ctx, ses, cb);
      }
      if (r === 1) {
        // The callback rides the box so it outlives the handshake.
        const box = { $: "TlsConn", ssl: ses, ctx: ctx, fd: fd, _cb: cb };
        return io_tup(listener, io_done(box));
      }
      let err = 0;
      try {
        err = lib.SSL_get_error(ses, r);
      } catch (e) {
        return fail(613, "handshake failed", fd, ctx, ses, cb);
      }
      if (err === 2) {
        io_park_on(fd, false, k, step);
        return undefined;
      }
      if (err === 3) {
        io_park_on(fd, true, k, step);
        return undefined;
      }
      return fail(613, "handshake failed", fd, ctx, ses, cb);
    };
    return step();
  };
  return go();
}

function tls_accept_raw_need() {
  return { read: true };
}

// The libssl symbols this effect needs, dlopened once per process and
// cached in a global slot (the dns_file style).
function tls_lib() {
  if (globalThis.BEND_TLS_ACC === undefined) {
    try {
      const ffi = require("bun:ffi");
      const T = { i: "i32", u: "u32", p: "ptr", v: "void" };
      const lib = ffi.dlopen("libssl.so.3", {
        TLS_server_method: { args: [], returns: T.p },
        SSL_CTX_new: { args: [T.p], returns: T.p },
        SSL_CTX_free: { args: [T.p], returns: T.v },
        SSL_CTX_use_certificate_file: { args: [T.p, T.p, T.i], returns: T.i },
        SSL_CTX_use_PrivateKey_file: { args: [T.p, T.p, T.i], returns: T.i },
        SSL_CTX_set_alpn_select_cb: { args: [T.p, T.p, T.p], returns: T.v },
        SSL_new: { args: [T.p], returns: T.p },
        SSL_free: { args: [T.p], returns: T.v },
        SSL_set_fd: { args: [T.p, T.i], returns: T.i },
        SSL_accept: { args: [T.p], returns: T.i },
        SSL_get_error: { args: [T.p, T.i], returns: T.i },
        SSL_select_next_proto: { args: [T.p, T.p, T.p, T.u, T.p, T.u], returns: T.i },
      });
      globalThis.BEND_TLS_ACC = {
        ...lib.symbols,
        ptr: ffi.ptr,
        JSCallback: ffi.JSCallback,
      };
    } catch (e) {
      globalThis.BEND_TLS_ACC = null;
    }
  }
  return globalThis.BEND_TLS_ACC;
}